client/json-rpc/src/client/async_client.rs

```diff
@@ -0,0 +1,1104 @@
+};
+pub use libra_json_rpc_types::response::JsonRpcResponse;
+
+use libra_crypto::hash::CryptoHash;
+use libra_types::{
+    account_address::AccountAddress,
+    transaction::{SignedTransaction, Transaction},
+};
+use serde::de::DeserializeOwned;
+use serde_json::json;
+use std::time::Duration;
+
+#[derive(Debug)]
+pub struct Client {
+    pub http_client: reqwest::Client,
+    pub server_url: reqwest::Url,
+    pub retry: std::boxed::Box<dyn RetryStrategy>,
+    pub last_known_state: std::sync::RwLock<Option<State>>,
+}
+
+impl Client {
+    pub fn new_from_url<T: reqwest::IntoUrl>(server_url: T) -> Result<Self, reqwest::Error> {
+        Ok(Self {
+            http_client: reqwest::ClientBuilder::new()
+                .use_native_tls()
+                .timeout(defaults::HTTP_REQUEST_TIMEOUT)
+                .build()
+                .expect("Unable to build Client."),
+            server_url: server_url.into_url()?,
+            last_known_state: std::sync::RwLock::new(None),
+            retry: std::boxed::Box::new(Retry {
+                max_retries: defaults::MAX_RETRIES,
+                delay: defaults::WAIT_DELAY,
+            }),
+        })
+    }
+
+    pub async fn wait_for_signed_transaction(
+        &self,
+        txn: &SignedTransaction,
+        timeout: Option<Duration>,
+        delay: Option<Duration>,
+    ) -> Result<jsonrpc::Transaction, WaitForTransactionError> {
+        self.wait_for_transaction(
+            &txn.sender(),
+            txn.sequence_number(),
+            txn.expiration_timestamp_secs(),
+            &Transaction::UserTransaction(txn.clone()).hash().to_hex(),
+            timeout,
+            delay,
+        )
+        .await
+    }
+
+    pub async fn wait_for_transaction(
+        &self,
+        address: &AccountAddress,
+        seq: u64,
+        expiration_time_secs: u64,
+        txn_hash: &str,
+        timeout: Option<Duration>,
+        delay: Option<Duration>,
+    ) -> Result<jsonrpc::Transaction, WaitForTransactionError> {
+        let start = std::time::Instant::now();
+        while start.elapsed() < timeout.unwrap_or(defaults::TIMEOUT) {
+            let account_txn = self
+                .get_account_transaction(address, seq, true)
+                .await
+                .map_err(WaitForTransactionError::GetTransactionError)?;
+            if let Some(txn) = account_txn {
+                if txn.hash != txn_hash {
+                    return Err(WaitForTransactionError::TransactionHashMismatchError(txn));
+                }
+                if let Some(ref vm_status) = txn.vm_status {
+                    if vm_status.r#type != jsonrpc::VM_STATUS_EXECUTED {
+                        return Err(WaitForTransactionError::TransactionExecutionFailed(txn));
+                    }
+                }
+                return Ok(txn);
+            }
+            if let Some(state) = self.last_known_state() {
+                if expiration_time_secs <= state.timestamp_usecs / 1_000_000 {
```

Comment:
I don't think there's any problem in writing this, my understanding is that integer division's only edge-case is division by 0, which clearly can't happen here. I created an issue for that here: https://github.com/rust-lang/rust-clippy/issues/6209